### PR TITLE
Output arrays in QASM 3 alias concatenations

### LINKED ISSUE #7214: Qasm3 Exporter needs to export arrays of qubtis for concatenation in alias of OpenQASM3.
### What is the expected enhancement?

Currently Qasm3 Exporter generates `qubit[2] _q; let q = _q[0] || _q[1];` to generate an array of `qubit`s in OpenQASM 3.
https://github.com/Qiskit/qiskit-terra/blob/main/test/python/circuit/test_circuit_qasm3.py#L34-L35

However, I think that the spec allows concatenating of only two arrays.
https://qiskit.github.io/openqasm/language/types.html#register-concatenation-and-slicing
 
In the above case, the type of ` _q[0]` and  `_q[1]` is `qubit` and not `qubit[1]`.

One of quick fixes may be generating `let q = _q[0:0] || _q[1:1];` instead of `let q = _q[0] || _q[1];`.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

The concatenation `||` operator in OpenQASM 3 is currently supposed to
operate only on arrays, but the exporter had been using it on single
qubits.  This changes the logic so that all the elements of a catenation
are arrays, and attempts to make each run as long as possible.

A new node to represent a `Range` was necessary to achieve this.  This
also meant reorganising the split between `IndexIdentifier`,
`IndexIdentifier2` and the builder-internal `indexIdentifierList`.
These names were originally taken from the OpenQASM 3 reference ANTLR
grammar, but the grammar has changed somewhat since then, and there are
also more descriptive names available.

(Note the internal Terra OpenQASM 3 AST is temporary, and due to be replaced after the 0.19 release.)

### Details and comments

Fix #7214

### CODE REVIEW

File: qiskit/qasm3/exporter.py [hunk 1]
```diff
@@ -448,15 +449,33 @@ def build_quantumdeclarations(self):
         )
         # aliases
         for qreg in self.circuit_ctx[-1].qregs:
-            qubits = []
+            elements = []
+            # Greedily consolidate runs of qubits into ranges.  We don't bother trying to handle
+            # steps; there's no need in generated code.  Even single qubits are referenced as a
```
Comment (kevinhartman) [inline qiskit/qasm3/exporter.py:454]: Is there a missing word at the end of this line?

Comment (jakelishman) [reply qiskit/qasm3/exporter.py]: Not even a full word, but 5/6 of one!  Thanks - fixed.

File: qiskit/qasm3/exporter.py [hunk 2]
```diff
@@ -448,15 +449,33 @@ def build_quantumdeclarations(self):
         )
         # aliases
```
Comment (kevinhartman) [inline qiskit/qasm3/exporter.py:450]: Not a request for changes, but just wondering if I'm understanding this correctly. Does this (very rough) pseudocode capture the idea?

```py
while qubit:
    range_bits = qubit.take(1).takeWhile(lambda bit: bit_location[bit] == bit_location[magic_prev_bit] + 1)
    qubit = qubit.skip(len(range_bits))
    elements.append(SI(range_bits))
```

Comment (jakelishman) [reply qiskit/qasm3/exporter.py:450]: Yeah, looks like it.  Most of the bookkeeping is just around setting up the first indices and ensuring that the final range is emitted too.

**Review by 1ucian0 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**